File: packages/cli/templates/scripts/emit-phase.js
#!/usr/bin/env bun
'use strict';
/**
 * emit-phase — record a `pipeline.phase` transition event from a SKILL.
 *
 * Problem this solves: `pipeline-phase.js` (PostToolUse hook) only emits
 * `pipeline.phase` when a `.claude/.pipeline-states/{spec}.json` file is
 * written and the phase changes. But ANALYZE runs in the parent context
 * BEFORE any pipeline-state file exists — so ANALYZE never produced
 * telemetry, and the dashboard showed `0` ANALYZE events forever.
 *
 * The SKILL is the only place that knows ANALYZE has started. This script
 * lets a SKILL emit the marker explicitly:
 *   bun .claude/scripts/emit-phase.js --spec add-login --to ANALYZE
 *
 * The emitted event is shape-identical to what `pipeline-phase.js` produces:
 *   event:   pipeline.phase
 *   payload: { from, to }
 *   spec:    {spec}
 * — so every downstream consumer (dashboard, metrics) treats both sources
 * uniformly.
 *
 * Idempotency: the script scans `.claude/.harness/events.jsonl` for the most
 * recent `pipeline.phase` event of the same spec. If its `to` already equals
 * the requested `--to` phase, the script silently skips the emit. This keeps
 * dedup logic in ONE place — SKILLs just call the script unconditionally.
 *
 * Cross-shell: no inline `bun -e` quoting. Fail-open: any internal error
 * exits 0 without emitting (telemetry must never break a pipeline).
 *
 * Exit codes:
 *   0  emitted, or idempotent skip, or fail-silent on internal error
 *   1  bad CLI arguments
 */

const fs = require('node:fs');
const path = require('node:path');

function parseArgs(argv) {
  const out = { spec: null, from: null, to: null };
  for (let i = 0; i < argv.length; i++) {
    const flag = argv[i];
    const next = argv[i + 1];
    switch (flag) {
      case '--spec':
        out.spec = next; i++; break;
      case '--from':
        out.from = next; i++; break;
      case '--to':
        out.to = next; i++; break;
      case '-h':
      case '--help':
        printHelp();
        process.exit(0);
        break;
      default:
        // ignore unknown flags rather than failing — fail-silent ethos
        break;
    }
  }
  return out;
}

function printHelp() {
  process.stdout.write(`emit-phase — record a pipeline.phase transition event.

Usage:
  bun emit-phase.js --spec <name> --to <PHASE> [--from <PHASE>]

  --spec NAME   spec identifier (required)
  --to PHASE    phase being entered, e.g. ANALYZE (required)
  --from PHASE  prior phase (optional; defaults to null)

Idempotent: skips the emit if the spec's latest pipeline.phase already
matches --to. Exit: 0 on emit/skip/silent-error, 1 on bad args.
`);
}

function resolveProjectDir() {
  if (process.env.CLAUDE_PROJECT_DIR) return process.env.CLAUDE_PROJECT_DIR;
  // Heuristic: script sits at .claude/scripts/, two levels up is project root.
  return path.resolve(__dirname, '..', '..');
}

function loadHarness(projectDir) {
  const harnessLib = path.join(projectDir, '.claude', 'hooks', '_lib', 'harness-event.js');
  if (!fs.existsSync(harnessLib)) return null;
  try {
    return require(harnessLib);
  } catch (_) {
    return null;
  }
}

/**
 * Returns the `to` phase of the most recent `pipeline.phase` event for the
 * given spec, or null if none found. Reads the events.jsonl tail-to-head so
 * the freshest record wins. Fail-soft: returns null on any error.
 */
function lastPhaseForSpec(eventsFile, spec) {
  try {
    if (!fs.existsSync(eventsFile)) return null;
    const lines = fs.readFileSync(eventsFile, 'utf8').split('\n');
    for (let i = lines.length - 1; i >= 0; i--) {
      const raw = lines[i].trim();
      if (!raw) continue;
      let obj;
      try { obj = JSON.parse(raw); } catch (_) { continue; }
      if (obj && obj.event === 'pipeline.phase' && obj.spec === spec) {
        return (obj.payload && obj.payload.to) || null;
      }
    }
  } catch (_) {}
  return null;
}

function main() {
  const args = parseArgs(process.argv.slice(2));

  if (!args.spec) {
    process.stderr.write('error: --spec required\n');
    printHelp();
    process.exit(1);
  }
  if (!args.to) {
    process.stderr.write('error: --to required\n');
    printHelp();
    process.exit(1);
  }

  const projectDir = resolveProjectDir();
  const harness = loadHarness(projectDir);
  if (!harness) {
    // Fail-silent: harness not installed yet. This is OK during bootstrap.
    process.exit(0);
  }

  // Idempotency: don't emit the same phase twice for one spec. If the spec's
  // latest pipeline.phase already lands on --to, this is a no-op.
  const eventsFile = harness.getEventsFile(projectDir);
  const last = lastPhaseForSpec(eventsFile, args.spec);
  if (last === args.to) {
    process.exit(0);
  }

  // `from` defaults to the spec's last known phase (null if none) — so a
  // SKILL only needs to pass --to. An explicit --from overrides this.
  const fromPhase = args.from || last || null;

  harness.emit('pipeline.phase', { from: fromPhase, to: args.to }, {
    cwd: projectDir,
    spec: args.spec,
    actor: { kind: 'orchestrator', id: 'emit-phase' },
  });
  process.exit(0);
}

main();

File: apps/rt/src/commands/event/emit_phase.rs
//! `mustard-rt run emit-phase` — a port of `scripts/emit-phase.js`.
//!
//! Records a `pipeline.phase` transition event from a SKILL. ANALYZE runs in
//! the parent context before any `pipeline-state` file exists, so the
//! `post_edit` pipeline-phase emitter never sees it — this is the only place
//! that knows ANALYZE started.
//!
//! The emitted event is shape-identical to what the `post_edit` module
//! produces (`event: "pipeline.phase"`, `payload: { from, to }`, `spec`), so
//! every downstream consumer treats both sources uniformly.
//!
//! Idempotency: the most recent `pipeline.phase` event for the same spec is
//! looked up from the per-spec NDJSON `.events/` dir; if its `to` already
//! equals the requested phase the emit is skipped. Events are written directly
//! to the NDJSON sink — no SQLite involved.

use crate::shared::context::{project_dir, session_id};
use crate::shared::events::writer_ndjson;
use mustard_core::time::now_iso8601;
use mustard_core::io::claude_paths::ClaudePaths;
use mustard_core::domain::model::event::{Actor, ActorKind, HarnessEvent, SCHEMA_VERSION};
use mustard_core::view::projection::read_harness_events_from_ndjson_dir;
use serde_json::json;
use std::path::Path;

/// Return the `to` phase of the most recent `pipeline.phase` event for `spec`
/// by reading the per-spec NDJSON `.events/` directory. Fail-open — a missing
/// events dir or any IO error yields `None` (the caller treats that as "phase
/// unknown").
///
/// This is the single source of truth for spec phase across the runtime; every
/// consumer that previously read `phaseName` from a pipeline-state JSON now
/// derives the phase through this helper instead.
#[must_use]
pub fn last_phase_for_spec(cwd: impl AsRef<Path>, spec: &str) -> Option<String> {
    let events_dir = ClaudePaths::for_project(cwd.as_ref())
        .and_then(|p| p.for_spec(spec))
        .ok()
        .map(|sp| sp.events_dir())
        .unwrap_or_else(|| {
            ClaudePaths::compose_unchecked(cwd.as_ref())
                .spec_dir()
                .join(spec)
                .join(".events")
        });
    let mut events = read_harness_events_from_ndjson_dir(&events_dir);
    events.sort_by(|a, b| a.ts.cmp(&b.ts));
    events
        .iter()
        .rev()
        .find(|e| e.event == "pipeline.phase" && e.spec.as_deref() == Some(spec))
        .and_then(|e| {
            e.payload
                .get("to")
                .and_then(|v| v.as_str())
                .map(str::to_string)
        })
}

/// What a phase emit did — the material of the ONE deterministic success line
/// [`run`] prints. `Recorded` carries the previous phase (`None` on a spec's
/// first transition); `AlreadyThere` is the idempotent short-circuit, which
/// must SAY so: printing nothing made "already in that phase" and "transition
/// recorded" indistinguishable from stdout.
#[derive(Debug, PartialEq, Eq)]
pub(crate) enum PhaseOutcome {
    /// The transition was written; `from` is the phase it left.
    Recorded { from: Option<String> },
    /// The spec's latest phase already equals the requested one — no write.
    AlreadyThere,
}

/// The one deterministic success line for a phase emit — the same
/// `{ok, kind, spec}` shape [`super::emit_pipeline`] echoes (added there
/// because that emitter used to succeed in TOTAL silence), extended with the
/// transition itself: `from` (null on a first transition), `to`, and
/// `idempotent: true` on the short-circuit. No timestamp/session — run
/// outputs are byte-compared in gates; the NDJSON row carries those.
fn success_line(outcome: &PhaseOutcome, spec: &str, to: &str) -> serde_json::Value {
    match outcome {
        PhaseOutcome::Recorded { from } => json!({
            "ok": true, "kind": "pipeline.phase", "spec": spec, "from": from, "to": to,
        }),
        PhaseOutcome::AlreadyThere => json!({
            "ok": true, "kind": "pipeline.phase", "spec": spec, "from": to, "to": to,
            "idempotent": true,
        }),
    }
}

/// Run `mustard-rt run emit-phase --spec <name> --to <PHASE> [--from <PHASE>]`.
///
/// Fail-open for telemetry: any internal failure (NDJSON write) degrades to a
/// no-op. On success it prints the deterministic confirmation line
/// ([`success_line`]) — previous phase and new, and the idempotent
/// short-circuit says so instead of staying silent. The **exception** is
/// `--to CLOSE`, which runs the close-gate sub-gates (debt/checklist/qa/build)
/// inline before writing the event. A strict gate failure prints the gate
/// reason on stderr, leaves the event un-written, and exits the process with
/// status `1` — same user-visible behavior as the legacy `close_gate` hook
/// that fired on a pipeline-state Write/Edit (the trigger that no longer
/// exists post-Wave 2).
pub fn run(spec: &str, to: &str, from: Option<&str>) {
    match run_at(Path::new(&project_dir()), spec, to, from) {
        Ok(outcome) => println!("{}", success_line(&outcome, spec, to)),
        Err(reason) => {
            eprintln!("{reason}");
            std::process::exit(1);
        }
    }
}

/// Cwd-aware miolo of [`run`]: record the `pipeline.phase` transition against
/// an explicit `cwd` instead of the process-global project dir. Reused by
/// `plan-materialize` (composing PLAN entry in-process) — which is why the
/// confirmation PRINT lives in the CLI entry [`run`], never here: a composite
/// caller's own stdout contract must not grow an interleaved line. `Ok`
/// carries the [`PhaseOutcome`] the caller may render; `Err` carries the
/// close-gate reason for a blocked `--to CLOSE` (the CLI entry turns that into
/// the legacy stderr + exit 1, composite callers fold it into their report).
pub(crate) fn run_at(
    cwd: &Path,
    spec: &str,
    to: &str,
    from: Option<&str>,
) -> Result<PhaseOutcome, String> {
    // Idempotency: skip when the spec's latest phase already lands on `to`.
    let last = last_phase_for_spec(cwd, spec);
    if last.as_deref() == Some(to) {
        return Ok(PhaseOutcome::AlreadyThere);
    }

    // CLOSE transition: run the close-gate sub-gates inline. A strict failure
    // blocks the transition; fail-open on any infrastructure error.
    if to.eq_ignore_ascii_case("CLOSE") {
        crate::commands::pipeline::close_gates::gate_close_for_spec(&cwd.to_string_lossy(), spec)?;
    }

    // `from` defaults to the spec's last known phase (null when none).
    let from_phase = from.map(str::to_string).or(last);

    // Wave-5 (project-profiler) write-back: when a spec transitions OUT of
    // EXECUTE, gather every concept-node id the resolver cached during the
    // session and write them back into `spec.md` as `injected` backlinks.
    // Fail-open — any IO error is swallowed, telemetry stays best-effort.

    let ts = now_iso8601();
    let sid = session_id();

    let event = HarnessEvent {
        v: SCHEMA_VERSION,
        ts: ts.clone(),
        session_id: sid.clone(),
        wave: 0,
        actor: Actor {
            kind: ActorKind::Orchestrator,
            id: Some("emit-phase".to_string()),
            actor_type: None,
        },
        event: "pipeline.phase".to_string(),
        payload: json!({ "from": from_phase.clone(), "to": to }),
        spec: Some(spec.to_string()),
    };

    // Write directly to the NDJSON sink. `pipeline.phase` was previously routed
    // to SQLite via `route::emit` (the `pipeline.*` prefix match), but
    // this sub-spec migrates all phase emission to the pure-NDJSON path.
    let kind = crate::shared::events::route::classify_kind(&event.event);
    let _ = writer_ndjson::write_event_with_ts(
        cwd,
        Some(spec),
        None,
        &sid,
        &event.event,
        kind,
        Some(event.wave),
        Some(&sid),
        Some("emit-phase"),
        None,
        &event.payload,
        Some(&ts),
    );
    Ok(PhaseOutcome::Recorded { from: from_phase })
}


#[cfg(test)]
mod tests {
    use super::*;
    use crate::shared::events::writer_ndjson::write_event;
    use serde_json::json;
    use tempfile::tempdir;

    fn emit_phase_event(project: &std::path::Path, spec: &str, to: &str) {
        let payload = json!({ "from": null, "to": to });
        let _ = write_event(
            project,
            Some(spec),
            None,
            "s",
            "pipeline.phase",
            "pipeline",
            Some(0),
            Some("s"),
            Some("emit-phase"),
            None,
            &payload,
        );
    }

    #[test]
    fn last_phase_reads_the_freshest_event() {
        let dir = tempdir().unwrap();
        emit_phase_event(dir.path(), "demo", "ANALYZE");
        emit_phase_event(dir.path(), "demo", "PLAN");
        assert_eq!(last_phase_for_spec(dir.path(), "demo").as_deref(), Some("PLAN"));
        assert_eq!(last_phase_for_spec(dir.path(), "other"), None);
    }

    #[test]
    fn last_phase_empty_log_is_none() {
        let dir = tempdir().unwrap();
        assert_eq!(last_phase_for_spec(dir.path(), "demo"), None);
    }

    /// AC-10: a recorded transition confirms itself — the success line names
    /// the previous phase and the new one — and the idempotent short-circuit
    /// SAYS so instead of staying silent (stdout used to be empty either way,
    /// making "already there" and "recorded" indistinguishable).
    #[test]
    fn emit_phase_confirms_the_transition() {
        let dir = tempdir().unwrap();

        // First transition: no previous phase — `from` is honest null.
        let first = run_at(dir.path(), "demo", "ANALYZE", None).expect("first emit");
        assert_eq!(first, PhaseOutcome::Recorded { from: None });
        let line = success_line(&first, "demo", "ANALYZE");
        assert_eq!(line["ok"], json!(true));
        assert_eq!(line["kind"], json!("pipeline.phase"));
        assert_eq!(line["spec"], json!("demo"));
        assert_eq!(line["from"], json!(null), "first transition has no previous phase: {line}");
        assert_eq!(line["to"], json!("ANALYZE"));
        assert!(line.get("idempotent").is_none(), "a real transition is not idempotent: {line}");

        // Second transition: the previous phase is read back and named.
        let second = run_at(dir.path(), "demo", "PLAN", None).expect("second emit");
        assert_eq!(second, PhaseOutcome::Recorded { from: Some("ANALYZE".to_string()) });
        let line = success_line(&second, "demo", "PLAN");
        assert_eq!(line["from"], json!("ANALYZE"), "previous phase named: {line}");
        assert_eq!(line["to"], json!("PLAN"));

        // Idempotent short-circuit: already-in-that-phase says so.
        let third = run_at(dir.path(), "demo", "PLAN", None).expect("idempotent emit");
        assert_eq!(third, PhaseOutcome::AlreadyThere);
        let line = success_line(&third, "demo", "PLAN");
        assert_eq!(line["idempotent"], json!(true), "the short-circuit is named: {line}");
        assert_eq!(line["from"], json!("PLAN"));
        assert_eq!(line["to"], json!("PLAN"));
    }
}
